consensus/types.go

```diff
@@ -1,53 +1,198 @@
+}
+
+type AggregatedSignature interface {
+	GetSignersList() *bitset.Bitset
+	GetSignature() *bls.Signature
+}
+
+type VoteMessage interface {
+	GetView() uint64
+	GetBlockHash() BlockHash
 
-// NewFastHotStuffConsensus creates a new persistent FastHotStuffConsensus object that internally
-// runs the Fast HotStuff consensus protocol, and signals the Server whenever it's ready to perform
-// an action.
+	GetPublicKey() *bls.PublicKey
+
+	// The validator's BLS signature of the (View, BlockHash) pair. This represents the validator's
+	// vote for this block. The block height is implicitly captured in the block hash.
+	GetSignature() *bls.Signature
+}
 
-type FastHotStuffConsensus struct {
-	nextBlockProposalTime time.Time
-	nextTimeoutTime       time.Time
+type TimeoutMessage interface {
+	GetView() uint64
+	GetHighQC() QuorumCertificate
 
-	internalTimersUpdated chan interface{}
-	quit                  chan interface{}
+	GetPublicKey() *bls.PublicKey
+	// The validator's BLS signature of the (View, HighQC.View) pair.
+	GetSignature() *bls.Signature
+}
 
-	// votesSeen and timeoutsSeen are in-memory maps of all the votes and timeouts we've
-	// seen so far, organized by their block hash and then by the public key of the voter.
-	votesSeen    map[BlockHash]map[bls.PublicKey]*bls.Signature // TODO: this should be a vote message
-	timeoutsSeen map[BlockHash]map[bls.PublicKey]*bls.Signature // TODO: this should be a timeout message
+type Block interface {
+	GetBlockHash() BlockHash
+	GetHeight() uint64
+	GetView() uint64
+	// This is a hybrid function that returns the QC from the block.
+	// - If the block is a normal block, this returns the QC from validators' votes for the previous block
+	// - If the block contains a timeout QC, this returns the validator high QC aggregated from
+	//   validators' timeout messages
+	// We are able to simplify the GetQC() to this behavior because this QC is only needed to construct
+	// a timeout QC for the next block in the event of a timeout. So, this QC will always be the latest QC
+	// at the current chain's tip that subsequent blocks will build on top of.
+	GetQC() QuorumCertificate
+}
 
-	ConsensusEvents chan *ConsensusEvent
+type BlockWithValidators struct {
+	Block Block
+	// The validator set for the next block height after the block. This validator set can be used to validate
+	// votes and timeouts used to build a QC that extends from the block. The validator set must be sorted
+	// in descending order of stake amount with a consistent tie breaking scheme.
+	Validators []Validator
 }
+
+// We want a large buffer for the signal channels to ensure threads don't block when trying to push new
+// signals.
+//
+// TODO: is a size of 100 enough? If we want to bullet-proof this, we could back it by a slice as a
+// secondary buffer. That seems unnecessary since every channel will only have signals pushed by a single
+// producer thread.
+const signalChannelBufferSize = 100
```

Comment:
100 is pretty small. I think the point of this is just to prevent blowing up memory. If each event is a kilobyte then you could easily do 10,000, which would use 10MB of memory at maximum.